File: app/src/main/java/foundation/e/blisslauncher/core/blur/BlurWallpaperProvider.java
package foundation.e.blisslauncher.core.blur;

import android.Manifest;
import android.app.WallpaperManager;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.support.v4.app.ActivityCompat;
import android.util.DisplayMetrics;
import android.view.Display;
import android.view.WindowManager;

import com.hoko.blur.HokoBlur;
import com.hoko.blur.processor.BlurProcessor;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import foundation.e.blisslauncher.BlissLauncher;
import foundation.e.blisslauncher.core.Utilities;

public class BlurWallpaperProvider {

    private final Context context;
    private final WallpaperManager wallpaperManager;

    private final BlurProcessor blurProcessor;

    private ExecutorService mDispatcher = Executors.newSingleThreadExecutor();

    private final Runnable updateRunnable = () -> updateWallpaper();

    private DisplayMetrics displayMetrics = new DisplayMetrics();

    private static final float SAMPLE_FACTOR = 8.0f;

    private static BlurWallpaperProvider sInstance;
    private Bitmap wallpaper;
    private volatile Future mFuture;
    private Listener listener;

    private BlurWallpaperProvider(Context context) {
        this.context = context;
        this.wallpaperManager = WallpaperManager.getInstance(context);
        blurProcessor = HokoBlur.with(context).sampleFactor(SAMPLE_FACTOR)
                .scheme(HokoBlur.SCHEME_OPENGL)
                .mode(HokoBlur.MODE_STACK)
                .forceCopy(false)
                .needUpscale(true)
                .processor();
        init();
        listener = (Listener) context;
    }

    public static BlurWallpaperProvider getInstance(Context context) {
        if (sInstance == null) {
            sInstance = new BlurWallpaperProvider(context);
        }
        return sInstance;
    }

    private void init() {
        updateWallpaper();
    }

    private void updateWallpaper() {
        if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
            return;
        }

        if (wallpaperManager.getWallpaperInfo() != null) {
            // Wallpaper is live wallpaper so can't support blur effect.
            return;
        }
        wallpaper = Utilities.drawableToBitmap(wallpaperManager.getDrawable(), true);
        wallpaper = scaleAndCropToScreenSize(wallpaper);
    }


    public void blur(int radius) {
        cancelPreTask(false);
        mFuture = mDispatcher.submit(new BlurTask(wallpaper, blurProcessor, radius) {
            @Override
            void onBlurSuccess(Bitmap bitmap) {
                if (bitmap != null && listener != null) {
                    listener.blurBackgroundLayer(bitmap);
                }
            }

            @Override
            void onBlurFailed(float factor) {
                listener.fallbackToDimBackground(factor);
            }
        });
    }

    public Bitmap mergeLauncherView(Bitmap launcherView) {
        updateWallpaper();
        if (wallpaper == null) {  // possibly we don't have access to read the wallpaper or the wallpaper is live wallpaper.
            return null;
        }
        int wallpaperWidth = wallpaper.getWidth();
        int wallpaperHeight = wallpaper.getHeight();
        int overlayWidth = launcherView.getWidth();
        int overlayHeight = launcherView.getHeight();

        // Hack for removing soft navigation bar
        if(overlayHeight > wallpaperHeight) {
            overlayHeight = wallpaperHeight;
            launcherView = Bitmap.createBitmap(launcherView, 0, 0, overlayWidth, overlayHeight);
        }

        float marginLeft = (float) (wallpaperWidth * 0.5 - overlayWidth * 0.5);
        float marginTop = (float) (wallpaperHeight * 0.5 - overlayHeight * 0.5);

        Bitmap finalBitmap = Bitmap.createBitmap(wallpaperWidth, wallpaperHeight, wallpaper.getConfig());
        Canvas canvas = new Canvas(finalBitmap);
        canvas.drawBitmap(wallpaper, new Matrix(), null);
        canvas.drawBitmap(launcherView, marginLeft, marginTop, null);
        return finalBitmap;
    }

    public void blurWithLauncherView(Bitmap view, int radius) {
        cancelPreTask(false);
        mFuture = mDispatcher.submit(new BlurTask(view, blurProcessor, radius) {
            @Override
            void onBlurSuccess(Bitmap bitmap) {
                if (bitmap != null && listener != null) {
                    listener.blurFrontLayer(bitmap);
                }
            }

            @Override
            void onBlurFailed(float factor) {
                listener.fallbackToDimBackground(factor);
            }
        });
    }

    public void cancelPreTask(boolean interrupt) {
        if (mFuture != null && !mFuture.isCancelled() && !mFuture.isDone()) {
            mFuture.cancel(interrupt);
            mFuture = null;
        }
    }

    private Bitmap scaleAndCropToScreenSize(Bitmap wallpaper) {
        WindowManager wm = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
        Display display = wm.getDefaultDisplay();
        display.getRealMetrics(displayMetrics);

        int width = displayMetrics.widthPixels;
        int height = displayMetrics.heightPixels;

        float widthFactor = ((float) width) / wallpaper.getWidth();
        float heightFactor = ((float) height) / wallpaper.getHeight();

        float upscaleFactor = Math.max(widthFactor, heightFactor);

        if (upscaleFactor <= 0) {
            return wallpaper;
        }

        int scaledWidth = (int) Math.max(width, (wallpaper.getWidth() * upscaleFactor));
        int scaledHeight = (int) Math.max(height, (wallpaper.getHeight() * upscaleFactor));

        wallpaper = Bitmap.createScaledBitmap(wallpaper, scaledWidth, scaledHeight, false);

        int navigationBarHeight = 0;
        if (BlissLauncher.getApplication(context).getDeviceProfile().hasSoftNavigationBar(context)) {

            int resourceId = context.getResources().getIdentifier("navigation_bar_height", "dimen", "android");
            if (resourceId > 0) {
                navigationBarHeight = context.getResources().getDimensionPixelSize(resourceId);
            }
        }

        int y;
        if (wallpaper.getHeight() > height) {
            y = (wallpaper.getHeight() - height) / 2;
        } else y = 0;

        return Bitmap.createBitmap(wallpaper, 0, y, width, height - navigationBarHeight);
    }

    public interface Listener {
        void blurBackgroundLayer(Bitmap bitmap);

        void blurFrontLayer(Bitmap bitmap);

        void fallbackToDimBackground(float dimAlpha);
    }

    public void clear() {
        listener = null;
        cancelPreTask(true);
        sInstance = null;
    }

    private abstract static class BlurTask implements Runnable {
        private Bitmap bitmap;
        private BlurProcessor blurProcessor;
        private int radius;

        BlurTask(Bitmap bitmap, BlurProcessor blurProcessor, int radius) {
            this.bitmap = bitmap;
            this.blurProcessor = blurProcessor;
            this.radius = radius;
        }

        @Override
        public void run() {
            if (bitmap != null && !bitmap.isRecycled() && blurProcessor != null) {
                blurProcessor.radius(radius);
                onBlurSuccess(blurProcessor.blur(bitmap));
            } else {
                onBlurFailed((float) radius / 15);
            }
        }

        abstract void onBlurSuccess(Bitmap bitmap);

        abstract void onBlurFailed(float factor);
    }
}

File: app/src/main/java/foundation/e/blisslauncher/core/blur/BlurWallpaperProvider.kt
package foundation.e.blisslauncher.core.blur

import android.Manifest
import android.app.WallpaperManager
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import android.util.DisplayMetrics
import android.util.Log
import android.view.WindowManager
import android.widget.Toast
import androidx.core.app.ActivityCompat
import foundation.e.blisslauncher.core.Utilities
import foundation.e.blisslauncher.core.runOnMainThread
import foundation.e.blisslauncher.core.safeForEach
import foundation.e.blisslauncher.core.utils.SingletonHolder
import foundation.e.blisslauncher.core.utils.ensureOnMainThread
import foundation.e.blisslauncher.core.utils.useApplicationContext
import java.util.ArrayList
import kotlin.math.max

class BlurWallpaperProvider(val context: Context) {

    private val wallpaperManager: WallpaperManager = WallpaperManager.getInstance(context)
    private val listeners = ArrayList<Listener>()
    private val displayMetrics = DisplayMetrics()

    var wallpaper: Bitmap? = null
        private set(value) {
            if (field != value) {
                field?.recycle()
                field = value
            }
        }
    var placeholder: Bitmap? = null
        private set(value) {
            if (field != value) {
                field?.recycle()
                field = value
            }
        }

    private val vibrancyPaint = Paint(Paint.FILTER_BITMAP_FLAG or Paint.ANTI_ALIAS_FLAG)

    private val mUpdateRunnable = Runnable { updateWallpaper() }

    private val wallpaperFilter = BlurWallpaperFilter(context)
    private var applyTask: WallpaperFilter.ApplyTask? = null

    private var updatePending = false

    init {
        isEnabled = getEnabledStatus()
        updateAsync()
    }

    private fun getEnabledStatus() = wallpaperManager.wallpaperInfo == null

    fun updateAsync() {
        Utilities.THREAD_POOL_EXECUTOR.execute(mUpdateRunnable)
    }

    private fun updateWallpaper() {
        if (applyTask != null) {
            updatePending = true
            return
        }

        // Prepare a placeholder before hand so that it can be used in case wallpaper is null
        val wm =
            context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val display = wm.defaultDisplay
        display.getRealMetrics(displayMetrics)
        val width = displayMetrics.widthPixels
        val height = displayMetrics.heightPixels
        placeholder = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(placeholder!!)
        canvas.drawColor(0x44000000)

        if (ActivityCompat.checkSelfPermission(
                context,
                Manifest.permission.READ_EXTERNAL_STORAGE
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            Log.d("BWP", "NO permission granted")
            return
        }

        val enabled = getEnabledStatus()
        if (enabled != isEnabled) {
            isEnabled = enabled
            runOnMainThread {
                listeners.safeForEach(Listener::onEnabledChanged)
            }
        }

        if (!isEnabled) {
            wallpaper = null
            return
        }

        var wallpaper = try {
            Utilities.drawableToBitmap(wallpaperManager.drawable, true) as Bitmap
        } catch (e: Exception) {
            runOnMainThread {
                val msg = "Failed: ${e.message}"
                Toast.makeText(context, msg, Toast.LENGTH_LONG).show()
                notifyWallpaperChanged()
            }
            return
        }
        wallpaper = scaleAndCropToScreenSize(wallpaper)
        wallpaper = applyVibrancy(wallpaper)
        applyTask = wallpaperFilter.apply(wallpaper).setCallback { result, error ->
            if (error == null) {
                this@BlurWallpaperProvider.wallpaper = result
                runOnMainThread(::notifyWallpaperChanged)
                wallpaper.recycle()
            } else {
                if (error is OutOfMemoryError) {
                    runOnMainThread {
                        Toast.makeText(context, "Failed", Toast.LENGTH_LONG).show()
                        notifyWallpaperChanged()
                    }
                }
                wallpaper.recycle()
            }
        }
        applyTask = null
        if (updatePending) {
            updatePending = false
            updateWallpaper()
        }
    }

    private fun notifyWallpaperChanged() {
        listeners.forEach(Listener::onWallpaperChanged)
    }

    private fun applyVibrancy(wallpaper: Bitmap?): Bitmap {
        val width = wallpaper!!.width
        val height = wallpaper.height
        val bitmap = Bitmap.createBitmap(
            width,
            height,
            Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas()
        canvas.setBitmap(bitmap)
        val colorMatrix = ColorMatrix()
        colorMatrix.setSaturation(1.25f)
        val filter = ColorMatrixColorFilter(colorMatrix)
        vibrancyPaint.colorFilter = filter
        canvas.drawBitmap(wallpaper, 0f, 0f, vibrancyPaint)
        wallpaper.recycle()
        return bitmap
    }

    private fun scaleAndCropToScreenSize(wallpaper: Bitmap): Bitmap {
        val wm =
            context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
        val display = wm.defaultDisplay
        display.getRealMetrics(displayMetrics)
        val width = displayMetrics.widthPixels
        val height = displayMetrics.heightPixels
        val widthFactor = width.toFloat() / wallpaper.width
        val heightFactor = height.toFloat() / wallpaper.height
        val upscaleFactor = Math.max(widthFactor, heightFactor)
        if (upscaleFactor <= 0) {
            return wallpaper
        }
        val scaledWidth =
            max(width.toFloat(), wallpaper.width * upscaleFactor).toInt()
        val scaledHeight =
            max(height.toFloat(), wallpaper.height * upscaleFactor).toInt()
        val scaledWallpaper =
            Bitmap.createScaledBitmap(wallpaper, scaledWidth, scaledHeight, false)
        val navigationBarHeight = 0
        /*if (BlissLauncher.getApplication(context).getDeviceProfile().hasSoftNavigationBar(context)) {

            int resourceId = context.getResources().getIdentifier("navigation_bar_height", "dimen", "android");
            if (resourceId > 0) {
                navigationBarHeight = context.getResources().getDimensionPixelSize(resourceId);
            }
        }*/
        val y: Int
        y = if (scaledWallpaper.height > height) {
            (scaledWallpaper.height - height) / 2
        } else 0

        return Bitmap.createBitmap(
            scaledWallpaper,
            0,
            y,
            width,
            height - navigationBarHeight
        )
    }

    fun addListener(listener: Listener) {
        listeners.add(listener)
    }

    fun removeListener(listener: Listener) {
        listeners.remove(listener)
    }

    fun createDrawable(): ShaderBlurDrawable {
        return ShaderBlurDrawable(this)
    }

    interface Listener {
        fun onWallpaperChanged() {}
        fun onEnabledChanged() {}
    }

    /*fun clear() {
        listener = null
        cancelPreTask(true)
        sInstance = null
    }*/

    companion object :
        SingletonHolder<BlurWallpaperProvider, Context>(ensureOnMainThread(useApplicationContext(::BlurWallpaperProvider))) {

        var isEnabled: Boolean = false
        private var sEnabledFlag: Int = 0

        fun isEnabled(flag: Int): Boolean {
            return isEnabled && sEnabledFlag and flag != 0
        }
    }
}
